be/src/exec/pipeline/scan/morsel.h

```diff
@@ -101,6 +104,48 @@ class LogicalSplitScanMorsel final : public ScanMorsel {
     std::vector<vectorized::ShortKeyRangeOptionPtr> _short_key_ranges;
 };
 
+/// MorselQueue
+class MorselQueueFactory {
+public:
+    virtual ~MorselQueueFactory() = default;
+
+    virtual MorselQueue* create(int driver_sequence) = 0;
+    virtual size_t size() const = 0;
+    virtual bool is_shared() const = 0;
+};
+
+class SharedMorselQueueFactory final : public MorselQueueFactory {
+public:
+    SharedMorselQueueFactory(MorselQueuePtr queue, int size) : _queue(std::move(queue)), _size(size) {}
+    ~SharedMorselQueueFactory() override = default;
+
+    MorselQueue* create(int driver_sequence) override { return _queue.get(); }
+    size_t size() const override { return _size; }
+    bool is_shared() const override { return true; }
+
+private:
+    MorselQueuePtr _queue;
+    const int _size;
+};
+
+class IndividualMorselQueueFactory final : public MorselQueueFactory {
```

Comment:
What's the meaning of Individual here?

Reply:
The morsel queues created from this factory don't share the same tablets source, that is, they are Individual for each other